### PR TITLE
Support dynamic shapes. Resolves #199

### LINKED ISSUE #199: Bypassing dynamic array limitation in JAX
Due to [limitations in JAX with dynamic array shape sizes](https://github.com/jax-ml/jax/discussions/24410), to interpolate a Fourier series to $Q$ non-uniform points with a type 2 nufft, one is often forced to pad the array `Q` to size `S` and instead interpolate to $|S| \sim |Q|^2$ points. So the cost becomes $O(K^d log(K^d) + S \log^d \epsilon^{-1})$ for the forward and similiar for the reverse autodiff gradient which further calls type 1 nuffts.

Given a boolean mask which denotes which points in the input array `S` are junk, would it be simple to add an option to finufft to skip the spreading computation to the junk points and just return some junk value at those indices? In generic python this would be as simple as [this](https://github.com/flatironinstitute/finufft/discussions/783#discussioncomment-15480295).

If this feature happens to be trivial to add, but would be considered bloat, I would be interested in adding this feature to a forked repository in the future, so it would still be useful to me to know that modification.

_Originally posted by @unalmis in https://github.com/flatironinstitute/finufft/discussions/783_

### LINKED ISSUE #1303: Differentiable code with dynamic shapes
A fundamental limitation of our auto diff tool JAX is it is [unable to handle dynamic size arrays](https://github.com/jax-ml/jax/discussions/18790) or jagged tensor operations. Both of these operations are supported by PyTorch, which I reference to better document what these operations are:
1. [Dynamic shapes](https://pytorch.org/docs/stable/torch.compiler_dynamic_shapes.html)
2. [Jagged tensor operations](https://pytorch.org/FBGEMM/fbgemm_gpu-overview/jagged-tensor-ops/JaggedTensorOps.html) 

So, unlike `c` code, in JAX we can't do `jnp.nonzero(a)[0]`, which constrains the algorithms we write. Sometimes it is possible to work around this by using a shape independent algorithm. For example, performing least squares regression to a variable number of points through equation 12 instead of 13: [least squares example](https://mathworld.wolfram.com/LeastSquaresFittingPolynomial.html). In other cases, it is not possible to write a shape independent algorithm without sacrificing performance.

It might be worthwile to implement a patch in `desc.backend` to work around JAX not being able to do 1. For example, we can make a differentiable `jnp.nonzero(a)[0]` with the following code

```python
import jax
import jax.numpy as jnp
import numpy as np

@jax.jit
def expensive_function(x):
    # pretend this is some non-trivial operation
    return jnp.dot(x, x) * x

grad_expensive_function = jnp.vectorize(jax.grad(expensive_function))

def nonzero_where(x):
    dummy = -1.0
    idx = np.nonzero(x)[0]
    y = jnp.full(x.shape, fill_value=dummy)
    y = y.at[idx].set(expensive_function(x[idx]))
    return y

def forward(x):
    return nonzero_where(x), x

def backward(x, gradient):
    idx = np.nonzero(x)[0]
    grad = jnp.zeros(x.shape)
    grad = grad.at[idx].set(grad_expensive_function(x[idx]) * gradient[idx])
    return (grad,)

fun = jax.custom_vjp(nonzero_where)
fun.defvjp(forward, backward)

x = jnp.array([1.0, 0.0, 3.0])
y = fun(x)

example_optimization_gradient = jax.grad(lambda x: fun(x).sum())
np.testing.assert_allclose(example_optimization_gradient(x), [3, 0, 27])
```

where the intention is to have a function in `desc.backend`  so that we can replace this logic
```python
dummy = -1.0
# this computes the expensive function for each element in x
y = jnp.where(x != 0, expensive_function(x), dummy) 
```
with 
```python
# this applies expensive function on nonzero elements only
y = nonzero_where(x, expensive_function)
```

Although this technique solves 1, the drawback is that JAX transformations, such as `jnp.vectorize` can not be applied to `nonzero_where` because it uses a numpy function. Hence to vectorize, one is limited to python for loops or list comprehension (which uses a c loop but still not ideal) instead of `vmap`, `map`, and `scan`. If the loop size is not large, this may not be that bad. (We already use list comprehension in other parts of code that is called when we compute things).

This may be a low-hanging fruit to implement to improve performance of objectives that use bounce integrals where the amount of  computation is proportional to the number of nonzero elements in an array (e.g. the number of wells).

### PR DESCRIPTION
**Pull request by unalmis**
## Summary
Prior to merging this pull request, the performance of `jax-finufft` is not competitive with `finufft`, pytorch `finufft`, or `julia`'s non-uniform ffts for use in algorithms where the interpolation points are not fixed a-priori.

This PR resolves #199 per the discussion in https://github.com/flatironinstitute/finufft/discussions/783#discussion-9323927 so that the performance of `jax-finufft` is now similar to that of `finufft`.

## Benchmarks
Here is a timing benchmark on my CPU for a small sized problem with `epsilon=1e-6` in a downstream application. Prior to this PR, every adjoint call to nufft1 took `>= 1 second` and the full computation was `34 seconds`. Now every adjoint call to nufft1 is `250 milliseconds`, and the full computation is `14 seconds`. These `improvements become larger` as the `sparsity` grows and error tolerance parameter for the nuffts epsilon tends to 0. Likewise, the improvement grows _linearly_ with the `problem size`. As this is called within an optimization loop where time and memory are tight, the improvement is significant.

### Before
Screenshot From 2026-03-29 15-22-56
Screenshot From 2026-03-29 15-22-36

### After
Screenshot From 2026-03-29 15-12-35
Screenshot From 2026-03-29 15-15-59

Note also that the downstream code has many tests which all pass with these modifications, and I have tested extensively.

### CODE REVIEW

**Comment by unalmis** [comment]
I do not have ability to request review to merge into main on this repository, so I'll just ping @lgarrison , @dfm in this comment.

**Comment by unalmis** [comment]
@lgarrison @dfm any chance to merge this yet?

**Comment by unalmis** [comment]
> any chance to merge this yet?

friendly reminder of previous message in case it was lost in GitHub @lgarrison or @dfm

**Comment by lgarrison** [comment]
Thanks @unalmis. I'll try to find time to review this, but our efforts are mostly spent on "keeping the lights on" rather than adding new features, so I can't promise a timeline. I appreciate that this is a big speedup for your application, though, so I'll try to give it a look.

**Comment by unalmis** [comment]
Thanks. I am happy to maintain this part of the code. It is unlikely that it will ever need that though considering most of the lines are adding an optional keyword argument to the function signature.
